compose_box: Add button to go the narrow message is being composed to.

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1243044

**Testing plan:** <!-- How have you tested? -->
Tested in the UI to see every case works.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
posted a gif at https://chat.zulip.org/#narrow/stream/137-feedback/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1247911. (Its size is bigger than what GitHub allows.)

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
Hmm, why doesn't this just call the `narrow.to_compose_target` function that we use for the `Ctrl+.` keyboard shortcut, which I think does the same thing?

chdinesh1089 commented:
When working on this, I didn't know about it and didn't look for something like `narrow.to_compose_target` as I'm not aware of `Ctrl+.` until you've mentioned it in the czo thread recently.

Will push a version that calls `narrow.to_compose_target`. The current version updates the button to be a link to topic/pm which can be right-clicked to copy the URL. I don't think it's super helpful but just wanted to know your thoughts before I push the other version.

chdinesh1089 commented:
Changed to use `narrow.to_compose_target`.

alya commented:
Super exciting to see progress on this! A couple of comments after playing around with the PR:

1. A blue box appears when I click on the button; we should prevent that from happening (`preventDefault`).
2. The button shows up when there is a stream but no topic entered. Instead, we should require both a stream and a topic name.

Thanks!

chdinesh1089 commented on static/styles/dark_theme.css:
```diff
@@ -519,6 +520,7 @@ body.dark-theme {
     }
 
     .expand_composebox_button:hover,
+    .narrow_to_compose_recipients:hover,
     .collapse_composebox_button:hover,
```
3. The color is the same as the others in the top row(expand, close). I wonder if it's because of the thickness that it looks different from the others?

chdinesh1089 commented:
1. It seems to be some default css thing with bootstrap. Added this to fix it for this button:
```
            &:focus,
            &:focus-visible {
                outline: none;
            }
```
2. Made the required changes.

Thanks for the feedback, Alya!

chdinesh1089 commented:
@alya Can you see if this looks good now?

alya commented:
Nice!

> Shall I add this for other buttons like expand compose, close compose too?

Yes, that would be great!

alya commented:
Let's add a "Go to conversation" tooltip to the button.

alya commented:
We should make sure the dark mode hover color is consistent with #21325 once that's merged.

chdinesh1089 commented:
> > Shall I add this for other buttons like expand compose, close compose too?
> 
> Yes, that would be great!

Added a commit on the top for this.

chdinesh1089 commented:
> Let's add a "Go to conversation" tooltip to the button. Otherwise, I think it's looking good!

oops, I've added the tooltip but didn't test after recent changes. There was some issue. Fixed it. Thanks for mentioning this!

The current wording is "Narrow to compose recipient (Ctrl+.)" (inspired from [Tim's czo message](https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1258319)). Is it okay? Do you think "Go to conversation" would be better? Perhaps 'narrow' sounds too technical? :thinking:

chdinesh1089 commented:
> We should make sure the dark mode hover color is consistent with #21325 once that's merged.

I've mentioned this in that PR too: https://github.com/zulip/zulip/pull/21325#issuecomment-1063147302
It would be a single line change to `color` property while the CSS selectors remain the same. So the color would match automatically on merging that if there's any change to the top buttons' color.

alya commented:
Hm, I guess @timabbott 's phrasing is more consistent with the `?` explanation ("Narrow to current compose box recipient"). "Go to conversation" sounds less technical, and might be clear enough given the context? @timabbott what do you think?

alya commented:
Including the keyboard shortcut on the tooltip is great, but a couple of comments:

1. It should be customized appropriately based on the operating system (cf. `?` menu, which correctly lists it as `CMD+.` on a Mac).
2. Could we add styling for it similar to the "scroll to bottom" tooltip?

![Screen Shot 2022-03-09 at 10 35 17 PM](https://user-images.githubusercontent.com/2090066/157603387-5c5bf192-afa6-4cf1-a741-231250dffc49.png)

Thanks!

chdinesh1089 commented:
Okay, made the relevant changes in a separate commit hoping it'd become easy for @timabbott to review.

Thanks for the feedback, @alya! These are all highly valuable in terms of UX.

It should look like this for you on mac:
![image](https://user-images.githubusercontent.com/44665669/157664890-792337c2-52cc-4c98-b0e0-d2c1f16f02c9.png)
For me it looks like 
![image](https://user-images.githubusercontent.com/44665669/157665029-e794e1fe-df7c-4cb8-81d1-17d569b97afa.png)

alya commented:
Cool, yeah, looking good (confirmed on dev)! @amanagr would you be up for doing a round of review on this PR?

alya commented:
Playing with it, I'm still thinking `Go to conversation (Ctrl + .)` will work better for the tooltip than the current phrasing, but we can let @timabbott comment before changing -- it's easy enough to adjust.

amanagr commented on static/styles/compose.css:
```diff
@@ -250,6 +250,11 @@
         &:hover {
             opacity: 1;
         }
+
```
Why do we need this change? We want outlines on `:focus` to help keyboard users.

chdinesh1089 replied:
Oh. We were getting the outlines on clicking too. I and @alya thought its a bug. Will remove this change then. Is there a way to hide it on clicks? Do we need an outline for clicks too?

chdinesh1089 replied:
Removed the commit adding this.

alya replied:
The compose box formatting buttons have outlines with the keyboard but not when you click on them, and I think that's the behavior we want.

amanagr replied:
Using `a` tag instead of `button` here should solve this.

chdinesh1089 replied:
Yes, that solved this, thanks!

amanagr commented on frontend_tests/node_tests/recent_topics.js:
```diff
@@ -100,6 +100,7 @@ mock_esm("../../static/js/muted_topics", {
 mock_esm("../../static/js/narrow", {
```
This commit doesn't make sense on its own. You can squash it to the commit which actually uses the function.

amanagr replied:
Yes, a commit with an empty function doesn't help.

chdinesh1089 replied:
Squashed them.

amanagr commented on static/assets/icons/arrow-left-circle.svg:
```diff
@@ -0,0 +1,3 @@
+<svg width="19" height="19" viewBox="0 0 19 19" fill="none" xmlns="http://www.w3.org/2000/svg">
```
We can use the fontawesome icon here - https://fontawesome.com/v4/icon/arrow-circle-o-left

chdinesh1089 replied:
We got it from @terpimost which we liked most from a [few variations](https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1316093). The fontawesome one looks a bit different. Do we want that?

alya replied:
Hm, could you post screenshots of the two versions (current icon and FontAwesome icon) next to each other?

chdinesh1089 replied:
Sure, FontAwesome followed by current:
![image](https://user-images.githubusercontent.com/44665669/158540820-30f889a8-4c90-459f-b91d-927122e44b19.png)
![image](https://user-images.githubusercontent.com/44665669/158540735-bd49de72-c354-4bbf-acee-713e2dcc2440.png)

alya replied:
Cool, thanks! I think the current icon looks better, so let's stick with that.

amanagr commented on static/styles/compose.css:
```diff
@@ -109,11 +109,28 @@
+                opacity: 1;
+            }
+
+            &:focus,
```
We want focus outline here for the icon.

alya replied:
Can we just have outlines with the keyboard but not on click?

amanagr replied:
The icon looks too glued to the topic input box in this iteration. We need more left margin. The 1px padding around the icon for the outline looks good.

chdinesh1089 replied:
Does this look good? 
![image](https://user-images.githubusercontent.com/44665669/158538571-22cbee98-881b-4380-b390-aa7ee63de92a.png)
added `margin: 3px`.
![image](https://user-images.githubusercontent.com/44665669/158538674-790bcc1e-d0c2-468b-9fdf-c5f23508db01.png)

chdinesh1089 replied:
@alya Yes, changed to that behavior. Thanks @amanagr, I didn't know changing to `a` tag fixes it. How does it work? Shouldn't we be able to achieve this behavior on anything with styling on `:focus-visible`? (It didn't work for button for me)

amanagr commented on static/templates/compose.hbs:
```diff
@@ -85,6 +85,7 @@
                                 <input type="text" class="recipient_box" name="stream_message_recipient_stream" id="stream_message_recipient_stream" maxlength="30" value="" placeholder="{{t 'Stream' }}" autocomplete="off" tabindex="0" aria-label="{{t 'Stream' }}" />
                                 <i class="fa fa-angle-right" aria-hidden="true"></i>
                                 <input type="text" class="recipient_box" name="stream_message_recipient_topic" id="stream_message_recipient_topic" maxlength="60" value="" placeholder="{{t 'Topic' }}" autocomplete="off" tabindex="0" aria-label="{{t 'Topic' }}" />
+                                <button type="button" class="narrow_to_compose_recipients zulip-icon zulip-icon-arrow-left-circle" aria-label="{{t 'Narrow to compose recipient (Ctrl+.)' }}" data-tippy-content="{{t 'Narrow to compose recipient (Ctrl+.)' }}"></button>
```
Can you fix the `tab` order? After pressing `Shift+tab` in the compose textarea, the focus should go to topic / pm input field.

amanagr replied:
The tab order looks correct now.

amanagr commented:
This doesn't work for PMs for me. Can you check?

amanagr reviewed (commented):
Thanks for working on this @chdinesh1089 , left a few comments.

amanagr commented on static/styles/compose.css:
```diff
@@ -109,11 +109,24 @@
+            background: transparent;
+            font-size: 18px;
+            padding: 1px;
+            line-height: 20px;
+            opacity: 0.6;
+            border: 0;
+
+            &:hover {
+                opacity: 1;
+            }
+        }
     }
 
     .pm_recipient {
         margin-left: 5px;
-        margin-right: 35px;
+        margin-right: 52px;
         display: flex;
```
We don't need this extra margin if we are not showing the button for pms.

chdinesh1089 replied:
Fixed. Forgot to remove this while removing previous pms specific changes.

chdinesh1089 commented:
Thanks for the reviews, @amanagr and @alya! I think I've addressed all the comments. Can you check now?

alya commented:
Looks good to me! @amanagr @timabbott is this ready to test-deploy on CZO?

amanagr commented on static/templates/compose.hbs:
```diff
@@ -82,6 +82,9 @@
                                     <i class="zulip-icon zulip-icon-globe" title="{{t 'This is a web-public stream' }}" aria-hidden="true"></i>
                                 </span>
+                                <a role="button" class="narrow_to_compose_recipients zulip-icon zulip-icon-arrow-left-circle order-1" data-tooltip-template-id="narrow_to_compose_recipients_tooltip" tabindex="0">
+                                    <template id="narrow_to_compose_recipients_tooltip">{{t 'Narrow to compose recipient' }} <span class="hotkey-hint">({{narrow_to_compose_recipients_key_html}})</span></template>
```
`Narrow to message recipient` would be better.

chdinesh1089 replied:
Ok, changed the tooltip wording to that.

amanagr commented:
Looks good to me!

chdinesh1089 commented:
Thanks so much, @amanagr and @alya, for the reviews!

@timabbott Could you review and perhaps test deploy on CZO?

timabbott commented on static/js/compose_actions.js:
```diff
@@ -136,6 +136,30 @@ export function expand_compose_box() {
+    const stream_name = compose_state.stream_name();
+    const topic = compose_state.topic() || undefined;
+    return narrow_state.stream() === stream_name && narrow_state.topic() === topic;
+}
```
This isn't a blocker for test-deploying this, but we should use `util.same_topic` or similar to do this check; topics and stream names are both case-insensitive.

chdinesh1089 replied:
Ok, changed to
```js
function composing_to_current_topic_narrow() {
    return (
        util.lower_same(compose_state.stream_name(), narrow_state.stream() || "") &&
        util.lower_same(compose_state.topic(), narrow_state.topic())
    );
}

timabbott commented:
It appears this doesn't show the button when composing to a private message conversation that doesn't match your current view; we should fix that.

chdinesh1089 commented:
> That is [intentional](https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1318548). We weren't sure where to show the icon.

That is [intentional](https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1318548). We weren't sure where to show the icon.

chdinesh1089 commented:
The color of the icon was defaulting to blue as described in https://chat.zulip.org/#narrow/stream/101-design/topic/narrow.20to.20topic.2Fpms.20when.20composing/near/1360246. Fixed it with `color: inherit` in CSS.

timabbott commented:
I'm deploying an updated version of this now.

timabbott commented:
Just adding a note here that I think we're waiting to merge this on adding a second line of text clarifying the situation when you're composing to a topic within the stream that one is narrowed to.

timabbott commented:
Merged, thanks @chdinesh1089!
